src/parquet/column_writer.cc

```diff
@@ -425,10 +430,20 @@ void ColumnWriter::WriteDataPage(const CompressedDataPage& page) {
 int64_t ColumnWriter::Close() {
   if (!closed_) {
     closed_ = true;
-    if (has_dictionary_ && !fallback_) {
+
+    if (has_dictionary_ && fallback_ && flush_on_close_) {
+      total_bytes_written_ += pager_->WriteDictionaryPage(saved_dictionary_page_[0]);
+      // Write all outstanding data to a new page
+      // Important that we buffer all pages for RowGroupWriter2
+      if (num_buffered_values_ > 0) {
+        AddDataPage();
+      }
+    } else if (has_dictionary_ && !fallback_) {
+      flush_on_close_ = false;
       WriteDictionaryPage();
     }
 
+    flush_on_close_ = false;
```

Comment:
This seems a tad hacky to me; I think there is a missing use of abstraction here which is the management of pages to be written to the file. The Parquet Java library has notions that provide for multiple implementations of this:

https://github.com/apache/parquet-mr/tree/master/parquet-column/src/main/java/org/apache/parquet/column/page

We don't need to necessarily do it in this patch, but using the existing PageWriter (or something similar if what is there now doesn't have the right features) abstraction with APIs to provide for

a) an implementation that accumulates pages in memory
b) an implementation that writes pages immediately

At some point we will want to make page writes asynchronous for better write performance so this refactoring will probably be desired at some point anyway

Reply:
The PageWriter class is responsible for writing compressed data pages out. But the goal of either writing compressed Pages or accumulating them is more efficient if its part of the `ColumnWriter`.